---
title: "Introduction to simplePHENOTYPES"
output:
  rmarkdown::html_document:
    toc: true
    toc_depth: 2
    number_sections: true
    theme: spacelab
    highlight: textmate
    df_print: tibble
author: Samuel B Fernandes and Alexander E Lipka
date: Sep 25, 2019

vignette: >
  %\VignetteIndexEntry{Introduction to simplePHENOTYPES}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#"
)
```
This short tutorial presents some of the possible genetic architectures that one could simulate, but it certainly does not explore all the possibilities. For more information on specific input parameters, please check the help documentation (?create_phenotypes).

# Installation
In order to install simplePHENOTYPES you will need to install SNPRelate and gdsfmt (both from Bioconductor), as well as mvtnorm, lqmm and data.table (from CRAN).
```{r installation, results = "hide", eval = FALSE}
#IMPORTANT 1[space]2 is setting CRAN and Bioconductor as repositories
setRepositories()
1 2
#Option 1: Installing from bitbucket:
##On Windows please uncomment and run the line below:
#options(download.file.method = "libcurl")
devtools::install_bitbucket("fernandessb/simplephenotypes", auth_user = "fernandessb", password = "RhtNaUHY66VdLESMRLCD", build_vignettes = TRUE)

#Option 2: Installing from source package:
install.packages("gdsfmt")
install.packages("SNPRelate")
install.packages("lqmm")
install.packages("data.table")
install.packages("mvtnorm")
# please replace [PATH] by your path to the simplePHENOTYPES_0.2.1.tar.gz file 
# or install it using RStudio's menu option to install from Package Arquive File.
install.packages("[PATH]/simplePHENOTYPES_0.2.1.tar.gz", repos = NULL, type = "source")
```

# Load sample dataset
Note that the example dataset is already numericalized. HapMap format may also be used with options  `genotypes_object`, `genotypes_file` or `genotypes_path`.
```{r load data}
library(simplePHENOTYPES)
data("SNP55K_maize282_maf04")
SNP55K_maize282_maf04[1:8, 1:10]
```

# Single Trait
Simulating ten replicates of a single trait with a heritability of 0.7. In this setting, the simulated trait is controlled by one large-effect QTN (additive effect of 0.9) and two small effect QTNs with additive effects following a geometric series starting with 0.2; thus the effect size of the first of these two QTNs is 0.2, and the effect size of the second is 0.2^2.
Please see help files (?create_phenotypes) to see which default values are being used.
```{r ST, eval = FALSE}
create_phenotypes(
  genotypes_object = SNP55K_maize282_maf04,
  additive_QTN_number = 3,
  additive_effect = 0.2,
  big_additive_QTN_effect = 0.9,
  rep = 10,
  h2 = 0.7)
```

# Multiple Traits: Pleiotropic Architecture
Simulating three traits `ntraits = 3` controlled by the same three additive QTN. One of these traits (Trait_1) is a "target", while the other is a correlated trait of lesser importance. The effect size of the largest-effect QTN is 0.3 for all traits, while the additive effect sizes are 0.04, 0.2 and 0.1 for each trait, respectively. Heritability for the target trait is 0.2, while the heritability of the two correlated traits is 0.4. Each replicate is being recorded in a different file (`output_format = "multi-file"`) in a folder named "Results_Pleotropic". The correlation between traits is not specified by the user; instead, the observed correlation is an artifact of different allelic effects for each trait. The same QTNs are used to generate phenotypes in all 10 replications (`rep_by = "experiment"`); alternatively, one could select different QTNs in each replicate using `rep_by = "QTN"` (default).
```{r MT P, results = "hide", eval = FALSE}
 test1 <-  create_phenotypes(
    genotypes_object = SNP55K_maize282_maf04,
    additive_QTN_number = 4,
    big_additive_QTN_effect = c(0.3, 0.3, 0.3),
    h2 = c(0.2,0.4, 0.4),
    additive_effect = c(0.04,0.2,0.1),
    ntraits = 3,
    rep = 10,
    rep_by = "experiment",
    output_format = "multi-file",
    architecture = "pleiotropic",
    output_dir = "Results_Pleiotropic",
    to_r = T,
    seed = 10
  )
  
 users_geometric <- matrix(c(0.04, 0.2, 0.1,
                            0.0016, 0.04, 0.01,
                            6.4e-05, 0.008, 0.001), nrow= 3, byrow = T)

 test2 <-  create_phenotypes(
   genotypes_object = SNP55K_maize282_maf04,
   additive_QTN_number = 4,
   big_additive_QTN_effect = c(0.3, 0.3, 0.3),
   h2 = c(0.2,0.4, 0.4),
   additive_effect = users_geometric,
   ntraits = 3,
   rep = 10,
   rep_by = "experiment",
   output_format = "multi-file",
   architecture = "pleiotropic",
   output_dir = "Results_Pleiotropic",
   to_r = T,
   sim_method = "user",
   seed = 10
 )
 
 all.equal(test1, test2)
```

# Multiple Traits: Partially Pleiotropic Architecture
Simulating 20 replicates of three traits, which are respectively controlled by seven, 13 and four QTNs. The first of the three pleiotropic QTNs (`overlap = 3`) will have a large additive QTN effect of 0.9. All other QTNs will have additive effects that follow a geometric series, where the effect size of the i^th QTN has an effect size of 0.3^i. Correlation among traits is assigned to be equal to the cor_matrix object. All 20 replicates of these three simulated traits will be saved in one file, specifically in a long format and with an additional column named "Rep". Results will be saved in a directory called "Results_Partially". In this example, the genotype file will be saved in numeric format. Additionally, all phenotypes will be assigned to an object called "sim_results" (to_r = TRUE) and loaded into R.
```{r MT PP, results = "hide", eval = FALSE}
cor_matrix <- matrix(c(   1, 0.4, -0.5,
                        0.4,   1,  0.7,
                       -0.5, 0.7,    1 ), 3)

sim_results <- create_phenotypes(
  genotypes_object = SNP55K_maize282_maf04,
  overlap = 3,
  specific_QTN_number = c(4, 10, 1),
  big_additive_QTN_effect = c(0.9, 0.9, 0.9),
  h2 = c(0.2, 0.4, 0.8),
  additive_effect = c(0.3, 0.3, 0.3),
  ntraits = 3,
  correlation = cor_matrix,
  rep = 20,
  output_dir = "Results_Partially",
  output_format = "long",
  architecture = "partially",
  out_geno = "numeric",
  to_r = TRUE
)
```


# Multiple Traits: Linkage Dissequilibrium Architecture
Simulating five replicates of two linked traits controlled by three additive QTNs. For each QTN, a marker is first selected, and then two separate markers that have an r^2 of at most 0.8 with the selected marker will be the respective QTNs of the two traits. The largest additive effect size of a pair of QTNs selected in this manner will be 0.1, and then the other QTN pairs will have additive effects that follow a geometric series, where the effect size of the i^th QTN is 0.02^i for one trait and 0.05^i for the other trait. Starting seed number is 200 and output phenotypes are saved in one file, but in a "wide" format with each replicate of two traits being added as additional columns. Plink fam, bim and bed files are also saved at Results_LD.

```{r MT LD, results = "hide", eval = FALSE}
create_phenotypes(
  genotypes_object = SNP55K_maize282_maf04,
  additive_QTN_number = 3,
  big_additive_QTN_effect = c(0.1, 0.1),
  h2 = c(0.2, 0.4),
  additive_effect = c(0.02, 0.05),
  rep = 5,
  seed = 200,
  output_format = "wide",
  architecture = "LD",
  output_dir = "Results_LD",
  out_geno = "plink",
  ld=0.8
)
```

# Multiple Traits: Partially Pleiotropic Architecture with Epistatic effects
Simulating 20 replicates of three traits. In each replicate, different SNPs are selected to be the QTNs for each trait. These traits are controlled by three pleiotropic additive QTNs (`overlap = 3`); two pleiotropic epistatic QTNs (`overlap_e = 2`); four, ten and one additive trait-specific QTN (`specific_QTN_number = c(4,10, 1)`);  and two, one and five epistatic trait-specific QTNs (`specific_e_QTN_number = c(2,1, 5)`). Results will be saved as ".fam" files used as gemma input; thus each replicate trait that is simulated will have its own unique output file.
```{r MT PP E, results = "hide", eval = FALSE}
create_phenotypes(
  genotypes_object = SNP55K_maize282_maf04,
  overlap = 3,
  overlap_e = 2,
  specific_QTN_number = c(4,10, 1),
  specific_e_QTN_number = c(2,1, 5),
  epistatic_effect = c(0.01, 0.4, 0.2),
  additive_effect = c(0.3, 0.2, 0.5),
  big_additive_QTN_effect = c(0.3, 0.2, 0.5),
  h2 = c(0.2,0.4, 0.8),
  ntraits = 3,
  rep = 20,
  rep_by = "QTN",
  output_dir = "Results_Partially_E",
  output_format = "gemma",
  architecture = "partially",
  model = "AE"
)
```

# Using your own data
If files are saved by chromosome, they can be read directly into create_phenotypes using options `genotypes_path` (recommendation: consider having all marker data files in a separate folder). If multiple files are saved in the same folder as the marker data, the parameter `shared_name` might be used to select only marker data. For example, if your data is saved as "WGS_chrm_1.hmp.txt", ..., "WGS_chrm_10.hmp.txt", one would use `shared_name = "WGS_chrm_"` .
```{r example, results = "hide", eval = FALSE}
create_phenotypes(
  genotypes_path = getwd(),
  shared_name = "WGS_chrm_",
  additive_QTN_number = 3,
  big_additive_QTN_effect = 0.1,
  h2 = 0.2,
  additive_effect = 0.02,
  rep = 5,
  seed = 200,
  output_format = "gemma",
  output_dir = "Results"
)
```

